### PR TITLE
feat: root-cause-aware finding grouping

### LINKED ISSUE #553: Root-cause-aware finding deduplication and grouping
## Problem

The 3-tier finding deduplication system in `FindingsRegistry` cannot detect that two findings with different titles share the same root cause.

**Example 1**: "Email Enumeration via Forgot Password Flow" and "Username Enumeration via Cognito Account Lockout Response" are both manifestations of **Cognito user-existence leaks**, but Tier 2 (title-stem matching) treats them as unrelated because their stems differ entirely.

**Example 2**: "Missing Rate Limiting on Email API" (CRITICAL 9.2) and "Unauthenticated Email Sending via SES API" (HIGH 7.7) exist on the **same endpoint** — the unauthenticated access is the root cause that makes rate limiting exploitable, yet they're reported as independent findings with inverted severity.

A professional pentester reviewing a generated report flagged both cases as findings that should be combined or explicitly linked.

## Cause

- **Tier 1** (exact match) only catches identical endpoint + vuln class
- **Tier 2** (app-wide) strips endpoint references from titles but can't detect semantic root-cause overlap between different vulnerability classes (e.g., `auth-bypass` + `missing-rate-limit` on same endpoint)
- **Tier 3** (semantic LLM) is conservative by design ("when in doubt, NOT a duplicate") and focuses on exact duplicates, not root-cause relationships
- There is no post-registration analysis pass that groups related findings

**Files**: `src/core/findings/registry.ts`

## Solution

1. Add a **post-registration `groupByRootCause()` method** on `FindingsRegistry` that runs after all findings are registered
2. Use LLM analysis to identify findings sharing a root cause: same endpoint with complementary vuln classes, or same underlying misconfiguration across different vectors
3. Annotate grouped findings with a `rootCauseGroup: string` identifier in the finding JSON
4. Extend the `Finding` type in `src/core/agents/offSecAgent/types.ts` with `rootCauseGroup?: string` and `relatedFindings?: string[]`
5. The report generator can then use root-cause groups to present related findings together and note the shared underlying issue

## Acceptance Criteria

- [ ] `groupByRootCause()` method identifies findings that share a root cause
- [ ] "Email Enumeration via Forgot Password" + "Username Enumeration via Cognito Lockout" are grouped
- [ ] "Unauthenticated Access" + "Missing Rate Limiting" on the same endpoint are grouped
- [ ] Grouped findings are annotated with `rootCauseGroup` identifier
- [ ] Unit tests cover both grouping scenarios

### PR DESCRIPTION
**Pull request by Yuvanesh-ux**
## Summary
- Add `groupByRootCause()` method to `FindingsRegistry` that uses LLM analysis to identify findings sharing a root cause
- Extend `Finding` type with optional `rootCauseGroup` and `relatedFindings` fields
- Groups findings like "Email Enumeration via Forgot Password" + "Username Enumeration via Cognito Lockout" that stem from the same underlying misconfiguration

Closes #553

## Test plan
- [x] Unit tests for grouping scenarios (Cognito enumeration, same-endpoint auth + rate-limiting)
- [x] Unit tests for edge cases (no model, fewer than 2 findings, unrelated findings)
- [x] Existing dedup tests still pass
- [x] Typecheck clean
- [x] Lint clean

---

> [!NOTE]
> **Medium Risk**
> Adds a new LLM-driven analysis step that mutates stored `Finding` objects with grouping metadata; incorrect model output or schema drift could mis-group issues and affect downstream reporting.
> 
> **Overview**
> Adds root-cause-aware grouping to `FindingsRegistry` via a new `groupByRootCause()` method that calls `generateObjectResponse` with a structured schema to return groups of finding indices and a `groupId`.
> 
> Extends the `Finding` schema with optional `rootCauseGroup` and `relatedFindings` fields, and adds unit tests covering positive grouping scenarios plus edge cases where grouping is skipped (no model / <2 findings) or returns no groups.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 13b1e27932c4b0b1962e9ffaae8a8545df76a6d5. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/core/findings/registry.ts [hunk 1]
```diff
@@ -511,6 +567,73 @@ export class FindingsRegistry {
...
+    if (!this.model || this.findings.length < 2) {
+      return [];
+    }
+
+    // Snapshot findings before the (potentially slow) LLM call so that
+    // concurrent register/unregister calls don't shift indices.
+    const snapshot = [...this.findings];
+    const prompt = buildRootCauseGroupingPrompt(snapshot);
+
+    const result = await generateObjectResponse({
```
Comment (jorgeraad) [inline src/core/findings/registry.ts]: Nit: We should catch potential errors either here or let it pass through, but then just add a @throws JSDoc on the function.

File: src/core/findings/registry.ts [hunk 2]
```diff
@@ -511,6 +567,73 @@ export class FindingsRegistry {
...
+   * Analyse all registered findings and group those that share a common
+   * root cause (e.g. same misconfiguration, one enables another).
+   *
+   * This is NOT deduplication — grouped findings are distinct issues that
+   * happen to stem from the same underlying weakness.
+   *
+   * Annotates findings in-place with `rootCauseGroup` and `relatedFindings`.
+   * Returns the groups array (empty when fewer than 2 findings or no model).
+   */
+  async groupByRootCause(): Promise<RootCauseGroup[]> {
```
Comment (jorgeraad) [inline src/core/findings/registry.ts:584]: Is `groupByRootCause` being called anywhere yet or will it be integrated in a follow-up? This PR marks the issue as closed.

**Review by jorgeraad — CHANGES_REQUESTED** [review]

**Comment by Yuvanesh-ux** [comment]
@jorgeraad Pls take a look now

**Review by jorgeraad — APPROVED** [review]
Looks good! Thanks for creating issues like #553 for all your PRs. Makes it easy to get the full context.
